pkg/connection/reconcile_geo_replication.go

```diff
@@ -0,0 +1,205 @@
+				r.conn.hasUnreadyResource = true
+				break
+			}
+		}
+	}
+
+	r.log.V(1).Info("Observe Done")
+	return nil
+}
+
+// Observe checks the updates of object
+func (r *PulsarGeoReplicationReconciler) Reconcile(ctx context.Context) error {
+	for i := range r.conn.geoReplications {
+		geoReplication := &r.conn.geoReplications[i]
+		if err := r.ReconcileGeoReplication(ctx, r.conn.pulsarAdmin, geoReplication); err != nil {
+			return fmt.Errorf("reconcile geo replication [%w]", err)
+		}
+	}
+	return nil
+}
+
+func (r *PulsarGeoReplicationReconciler) ReconcileGeoReplication(ctx context.Context, pulsarAdmin admin.PulsarAdmin,
+	geoReplication *resourcev1alpha1.PulsarGeoReplication) error {
+
+	log := r.log.WithValues("pulsargeoreplication", geoReplication.Name, "namespace", geoReplication.Namespace)
+	log.V(1).Info("Start Reconcile")
+
+	destConnection := &resourcev1alpha1.PulsarConnection{}
+	namespacedName := types.NamespacedName{
+		Name:      geoReplication.Spec.DestinationConnectionRef.Name,
+		Namespace: geoReplication.Namespace,
+	}
+	if err := r.conn.client.Get(ctx, namespacedName, destConnection); err != nil {
+		log.Error(err, "Failed to get destination connection for geo replication")
+		if apierrors.IsNotFound(err) {
+			return err
+		}
+	}
+	destClusterName := destConnection.Spec.ClusterName
+	if len(destClusterName) == 0 {
+		return fmt.Errorf("ClusterName is empty in destination connection")
+	}
+
+	if !geoReplication.DeletionTimestamp.IsZero() {
+		log.Info("Deleting GeoReplication", "LifecyclePolicy", geoReplication.Spec.LifecyclePolicy)
+		skip := false
+		for _, i := range r.conn.tenants {
+			if len(i.Spec.GeoReplicationRefs) != 0 {
+				for _, j := range i.Spec.GeoReplicationRefs {
+					if j.Name == geoReplication.Name {
+						skip = true
+						log.Info("There is still a tenant used this geo replication. The GeoReplication couldn't be deleted", "tenant", i.Name)
+					}
+				}
+			}
+		}
+		if !skip {
+			if geoReplication.Spec.LifecyclePolicy == resourcev1alpha1.CleanUpAfterDeletion {
+				// Delete the cluster that created with destination cluster info.
+				// TODO it can only be deleted after the cluster has been removed from the tenant, namespace, and topic
+				if err := pulsarAdmin.DeleteCluster(destClusterName); err != nil && admin.IsNotFound(err) {
+					log.Error(err, "Failed to delete geo replication cluster")
+					return err
+				}
+			}
+			controllerutil.RemoveFinalizer(geoReplication, resourcev1alpha1.FinalizerName)
+			if err := r.conn.client.Update(ctx, geoReplication); err != nil {
+				log.Error(err, "Failed to remove finalizer")
+				return err
+			}
+		}
+	}
+	controllerutil.AddFinalizer(geoReplication, resourcev1alpha1.FinalizerName)
+	if err := r.conn.client.Update(ctx, geoReplication); err != nil {
+		log.Error(err, "Failed to remove finalizer")
+		return err
+	}
+
+	if resourcev1alpha1.IsPulsarResourceReady(geoReplication) {
+		log.V(1).Info("Resource is ready")
+		return nil
```

Comment:
Would you mind leaving a comment here as to why we're returning when the pulsar resource is ready?  I'm not familiar enough with this to know why the rest is skipped when we hit this point.  Is it because this means a previous reconcile setup the clusterParam successfully on line 173?